src/main/java/cubicchunks/worldgen/gui/component/UIBlockStateButton.java

```diff
@@ -0,0 +1,113 @@
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.GlStateManager;
+import net.minecraft.client.renderer.Tessellator;
+import net.minecraft.client.renderer.VertexBuffer;
+import net.minecraft.client.renderer.texture.ITextureObject;
+import net.minecraft.client.renderer.texture.TextureMap;
+import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
+import net.minecraft.client.renderer.vertex.VertexFormat;
+import net.minecraft.init.SoundEvents;
+import net.minecraft.util.math.BlockPos;
+
+public class UIBlockStateButton extends UIComponent<UIBlockStateButton> {
+
+    private IBlockState iBlockState;
+    private final Consumer<IBlockState> onClick;
+
+    public UIBlockStateButton(MalisisGui gui, IBlockState iBlockState1, Consumer<IBlockState> clickAction) {
+        super(gui);
+        iBlockState = iBlockState1;
+        onClick = clickAction;
+        setTooltip(generateTooltip(iBlockState));
+        setSize(24, 24);
+    }
+
+    private String generateTooltip(IBlockState blockState) {
+        StringBuffer sb = new StringBuffer(128);
+        sb.append(blockState.getBlock().getLocalizedName());
+        for (Entry<IProperty<?>, Comparable<?>> entry : blockState.getProperties().entrySet()) {
+            sb.append(" \n ");
+            sb.append(entry.getKey().getName());
+            sb.append(" = ");
+            sb.append(entry.getValue().toString());
+        }
+        return sb.toString();
+    }
+
+    public void setBlockState(IBlockState iBlockState1) {
+        iBlockState = iBlockState1;
+        setTooltip(generateTooltip(iBlockState));
+    }
+
+    @Override
+    public boolean onClick(int x, int y) {
+        MalisisGui.playSound(SoundEvents.UI_BUTTON_CLICK);
+        onClick.accept(iBlockState);
+        return true;
+    }
+
+    @Override
+    public void drawForeground(GuiRenderer renderer, int mouseX, int mouseY, float partialTick) {
+        if (iBlockState != null) {
+            this.drawBlock(renderer);
+        }
+    }
+
+    private void drawBlock(GuiRenderer renderer) {
+
+        VertexBuffer vertexbuffer = Tessellator.getInstance().getBuffer();
+        Tessellator.getInstance().draw();
+        ITextureObject blockTexture = Minecraft.getMinecraft().getTextureManager().getTexture(TextureMap.LOCATION_BLOCKS_TEXTURE);
+        GL11.glBindTexture(GL11.GL_TEXTURE_2D, blockTexture.getGlTextureId());
+        VertexFormat format = DefaultVertexFormats.BLOCK;
+        GlStateManager.pushMatrix();
+        GlStateManager.translate((float) this.screenX(), (float) this.screenY() + 16f, 100.0F);
+        GlStateManager.scale(16.0F, -16.0F, 16.0F);
+        GlStateManager.rotate(45f, 1f, 1f, 0f);
+        vertexbuffer.begin(GL11.GL_QUADS, format);
+        Minecraft.getMinecraft().getBlockRendererDispatcher().renderBlock(iBlockState, new BlockPos(0, 0, 0), DummyWorld.getInstance(), vertexbuffer);
```

Comment:
Might be better if the dummy world returned the blockstate you want to draw at position (0, 0, 0). I'm not sure exactly how it's used but it may confuse some mods if you don't do that.